### PR TITLE
OpenCL: Default to table based AES, now backed in local memory

### LINKED ISSUE #5594: Faster (by a magnitude) OpenCL AES code
https://github.com/openwall/john/pull/5591#issuecomment-2507661775

The default KeePass benchmark includes old low iteration test vectors. To see modern, real-life cost, edit opencl_keepass_fmt_plug.c, change the define of KEEPASS_REAL_COST_TEST_VECTORS to 1, rebuild and benchmark with `-cost=31250000` (it will take a looong time).

Each KDF-AES round is 32 bytes of AES-256 so at this cost it will encrypt exactly 1 GB with AES-256 per candidate - great for testing AES performance. Our shared OpenCL AES comes in two flavors. Using the "bitsliced AES" (default for GPU - I kinda doubt bitsliced is a correct description but it does _some_ things in parallel), Super's 1080 does 20.5 GB/s which I believe [could be improved](https://eprint.iacr.org/2021/646) with <s>at least 10x</s> over 5x, (and forcing it to run our older, even worse, vanilla AES it's down to 1.15 GB/s which is essentially useless for anything but decrypting a short verifier).

The faster GPU code (CUDA) is at https://www.github.com/cihangirtezcan/CUDA_AES and there's also a [freely available PDF](https://eprint.iacr.org/2021/646.pdf).

### PR DESCRIPTION
**Pull request by magnumripper**
This revised version pushes 1.4 Tbps of AES-128 decryption (axcrypt) or 914 Gbps of AES-256 encryption (keepass) on a 4070ti.

The bitsliced code we defaulted to before is really good but it's register hungry. It has some merits when two or more blocks are encrypted/decrypted at once (does two in parallel) but still is slower than table based now. We can still opt in to use it.

Closes #5594

### CODE REVIEW

**Comment by magnumripper** [comment]
Up to 6.6x boost seen on super's AMD, and a mere 3x on the nvidias. I was hoping to achieve 10x but it turns out the bitsliced AES we used was too good as baseline for that to happen :) Anyway we seem to be on par with hashcat now.

**Review by solardiz — APPROVED** [review]
Cool stuff, but impossible to review for real without diving into it.

As a minor suggestion, maybe the moving of tables to a separate file can be a separate commit?

**Comment by magnumripper** [comment]
> impossible to review for real without diving into it.

The important changes are to opencl_aes_plain.h and are surprisingly few due to macros.

> As a minor suggestion, maybe the moving of tables to a separate file can be a separate commit?

I fail to see the point of that, but did so now. BTW the bitlocker format was also changed to use that table file but was otherwise not changed - it has its own copy of more or less identical (afaics) table based AES but did not gain anything from using local memory as the AES part of it is insignificant. So I did not commit any such changes to it.

File: src/opencl_keepass_fmt_plug.c [hunk 1]
```diff
@@ -53,10 +53,13 @@ typedef struct {
 	uint32_t cracked;
 } result;
 
+#define AES_MAXNR   14
```
Comment (solardiz) [inline src/opencl_keepass_fmt_plug.c:56]: FWIW, this addition of `AES_MAXNR` to `opencl_keepass_fmt_plug.c` looks unused.

**Comment by solardiz** [comment]
> > As a minor suggestion, maybe the moving of tables to a separate file can be a separate commit?
> 
> I fail to see the point of that, but did so now.

Thank you for doing it. Looks cleaner to me that way, and makes the actual changes (in other commits) stand out.

File: src/opencl_keepass_fmt_plug.c (hunk 1 above)
Comment (magnumripper) [reply src/opencl_keepass_fmt_plug.c:56]: Good catch, that's a remnant from the older state struct.

File: run/opencl/7z_kernel.cl [hunk 2]
```diff
@@ -162,7 +163,7 @@ __kernel void sevenzip_aes(__constant sevenzip_salt *salt,
 	/* Early rejection if possible (only decrypt last 16 bytes) */
 	if (pad > 0 && salt->length >= 32) {
 		uint8_t buf[16];
-		AES_KEY akey;
+		AES_KEY akey; akey.lt = &lt;
```
Comment (magnumripper) [inline run/opencl/7z_kernel.cl:166]: Oh BTW I first had this as `AES_KEY akey = { .lt = &lt };` everywhere, but some device did not like that. Unfortunately I forget which. Good to know, perhaps I should start a wiki page listing knowledge like that. Another example would be the `static` vs `inline` vs `static inline` that we had to add a [workaround for in opencl_misc.h](https://github.com/openwall/john/blob/18eae8da9d095bd4909d3a1f32294c9d0d4bfa08/run/opencl/opencl_misc.h#L54-L72).

**Comment by Br0kenUK** [comment]
Hi, can you use inverse sbox to speed up the invert key mix columns. I got 40% speed up doing it in hashcats code on aes256

**Comment by magnumripper** [comment]
> Hi, can you use inverse sbox to speed up the invert key. I got 40% speed up doing it in hashcats code on aes256

Are we not doing that already?

**Comment by Br0kenUK** [comment]
> > Hi, can you use inverse sbox to speed up the invert key. I got 40% speed up doing it in hashcats code on aes256
> 
> Are we not doing that already?

I don't think so? Especially if your speed is only on par with hashcat

**Comment by magnumripper** [comment]
I'm trying to context switch into this. My current guess is you mean we don't have the four T-tables for decryption, only a plain reverse sbox table - right?

Is your code in hashcat already or are you planning a PR? Feel free to add a PR here as well!

**Comment by Br0kenUK** [comment]
> > Hi, can you use inverse sbox to speed up the invert key. I got 40% speed up doing it in hashcats code on aes256
> 
> Are we not doing that already?

https://pastebin.com/b9CBfXRS

Above is roughly what I'm doing, I have left some random things out unrelated like hashcats decrypt function. This was mostly chatgpts idea so take it with a grain of salt, regardless it got me 40% increase in speed on a certain hashcat mode which also has additional md5 steps( and it works perfect). The only thing that has actually changed is the invert key step and the tables. And no I havn't PR'd it there, from what I understand they dont care much about improving AES.

**Comment by magnumripper** [comment]
You are very welcome to make a PR for us, for fun and fame 😉 

If not, I will look into it! And anyway thanks a lot for this suggestion!

**Comment by Br0kenUK** [comment]
I do not care for fame on things that was not my idea. I just randomly came across this issue and suggested it as it helped me. I won't be PRing it as I dont have the time to figure out your code aswell. GL.

**Pull request merged**